File: fm-database-web/src/lib/fmdb/somatic-read.ts
/**
 * The chief-complaint read, for the CLIENT app.
 *
 * A TS mirror of `fmdb/assess/somatic_read.py`. The coach surface calls that
 * Python through a shim; the client app cannot — `client-app.ts` never shells
 * out, it reads the catalogue off disk — so the matching is duplicated here on
 * purpose, the same way `lab-nutrient-priorities.ts` mirrors its Python.
 * KEEP THE TWO IN LOCKSTEP; `somatic-read.test.ts` pins this against output
 * captured from the real Python on real client records.
 *
 * WHY IT MUST BE A MIRROR AND NOT A REWRITE. Matching a free-text condition is
 * subtler than it looks, and one shortcut is already ruled out: resolving only
 * the slugs that somatic maps target, rather than the whole catalogue. Eleven
 * map targets — `hypertension`, `migraine`, `asthma`, `palpitations` among them
 * — are ALSO aliases of some other entity, so a reduced index resolves them
 * differently from the full one. Building the full index costs 42ms for 947
 * files, once per process. Fidelity is cheaper than the bug.
 *
 * The gate is elsewhere. This module answers "what does the book say about
 * what this client came in for"; it does NOT decide what the client may see.
 * That is `client-app.ts`, against the map's sensitivity and the client's own
 * `mind_body_depth`.
 */

import fs from "node:fs";
import path from "node:path";

import yaml from "js-yaml";

import { getCataloguePath } from "./paths";

type Dict = Record<string, unknown>;

const asStr = (v: unknown): string => (typeof v === "string" ? v : "");
const asStrArr = (v: unknown): string[] =>
  Array.isArray(v) ? v.map(asStr).filter(Boolean) : [];

/** Words that carry no clinical meaning when matching a condition string. */
const NOISE = new Set([
  "suspected", "confirmed", "possible", "probable", "mild", "moderate", "severe",
  "chronic", "acute", "on", "off", "treatment", "previously", "unreported",
  "and", "or", "the", "a", "of", "with", "in", "at", "to", "for", "non", "now",
  "history", "past", "current", "ongoing", "recurrent", "grade", "type",
]);

/**
 * Candidate lookup keys from one free-text condition, longest first.
 *
 * Real entries look like "Hypertension — ON TREATMENT (previously unreported)
 * — Telma 40 twice daily". The clinical name is almost always the head, before
 * the first dash, comma, colon or bracket.
 */
export function conditionPhrases(condition: string): string[] {
  const head = String(condition).split(/[—–\-(,:;/]/)[0].replace(/\s+/g, " ").trim().toLowerCase();
  if (!head) return [];
  const words = (head.match(/[a-z0-9']+/g) ?? []).filter((w) => !NOISE.has(w));
  const out: string[] = [];
  if (words.length) {
    out.push(words.join("-"));                       // full head
    for (const n of [3, 2]) {                        // leading n-grams
      if (words.length > n) out.push(words.slice(0, n).join("-"));
    }
    if (words.length > 1) out.push(words[words.length - 1]);  // trailing noun
    out.push(words[0]);
  }
  // de-dup, keep order, drop 1-2 char keys which match noisily
  const seen = new Set<string>();
  const keys: string[] = [];
  for (const k of out) {
    if (k.length > 2 && !seen.has(k)) {
      seen.add(k);
      keys.push(k);
    }
  }
  return keys;
}

export interface SomaticRoot {
  pattern: string;
  note: string;
}

export interface ChiefRead {
  /** the condition exactly as the coach wrote it */
  condition: string;
  targetSlug: string;
  mapSlug: string;
  displayName: string;
  sensitivity: string;
  /** coach_only_note is set — never auto-surface, whatever the depth */
  gated: boolean;
  roots: SomaticRoot[];
  reframe: string;
  inquiryQuestion: string;
  somaticPractice: string;
  differentialNote: string;
}

/** Safe to surface unsupervised, at the client's `full` depth. */
export function isClientSafe(r: ChiefRead): boolean {
  return r.sensitivity === "general" && !r.gated;
}

/* ---- catalogue indexes, built once per process ------------------------ */

function readDir(dir: string): Dict[] {
  let names: string[];
  try {
    names = fs.readdirSync(dir).filter((n) => n.endsWith(".yaml")).sort();
  } catch {
    return [];
  }
  const out: Dict[] = [];
  for (const n of names) {
    try {
      const d = yaml.load(fs.readFileSync(path.join(dir, n), "utf8"));
      if (d && typeof d === "object") out.push(d as Dict);
    } catch {
      /* a malformed entry must not take the whole read down */
    }
  }
  return out;
}

/**
 * {slug-or-alias → canonical-slug}. Mirrors `validator._resolve_index`,
 * including its conflict policy: a canonical slug always wins over an alias
 * that would shadow it.
 */
function resolveIndex(items: Dict[]): Map<string, string> {
  const index = new Map<string, string>();
  for (const it of items) {
    const canonical = asStr(it.slug);
    if (!canonical) continue;
    index.set(canonical, canonical);
  }
  for (const it of items) {
    const canonical = asStr(it.slug);
    if (!canonical) continue;
    for (const alias of asStrArr(it.aliases)) {
      // Don't let an alias shadow another entity's canonical slug
      if (!index.has(alias) || index.get(alias) !== alias) index.set(alias, canonical);
    }
  }
  return index;
}

interface Catalogue {
  resolve: (s: string) => string;
  byTarget: Map<string, Dict>;
}

let CACHE: Catalogue | null = null;

/** Exposed for tests, which need a clean build after changing the fixture root. */
export function _resetSomaticCache(): void {
  CACHE = null;
}

function catalogue(): Catalogue {
  if (CACHE) return CACHE;
  const root = getCataloguePath();
  const sym = resolveIndex(readDir(path.join(root, "symptoms")));
  const top = resolveIndex(readDir(path.join(root, "topics")));
  // symptoms first, then topics — the same precedence somatic-read.py uses
  const resolve = (s: string) => sym.get(s) ?? top.get(s) ?? s;

  const byTarget = new Map<string, Dict>();
  for (const m of readDir(path.join(root, "somatic_maps"))) {
    // the TARGET is resolved too: a map on `hypertension` lands under
    // `high-blood-pressure`, which is what a client's condition resolves to
    const key = resolve(asStr(m.target_slug));
    if (!byTarget.has(key)) byTarget.set(key, m);
  }
  CACHE = { resolve, byTarget };
  return CACHE;
}

/**
 * Map a client's stated conditions onto catalogue somatic maps.
 *
 * Ordered as the coach wrote the conditions: the first-listed condition is
 * almost always the presenting one.
 */
export function readChiefComplaints(conditions: string[], limit?: number): ChiefRead[] {
  const { resolve, byTarget } = catalogue();
  const out: ChiefRead[] = [];
  const seen = new Set<string>();

  for (const cond of conditions) {
    for (const key of conditionPhrases(cond)) {
      const canon = resolve(key);
      const m = byTarget.get(canon);
      const mapSlug = m ? asStr(m.slug) : "";
      if (!m || seen.has(mapSlug)) continue;
      seen.add(mapSlug);
      out.push({
        condition: String(cond),
        targetSlug: canon,
        mapSlug,
        displayName: asStr(m.display_name),
        sensitivity: asStr(m.sensitivity) || "sensitive",
        gated: asStr(m.coach_only_note).trim().length > 0,
        roots: (Array.isArray(m.emotional_roots) ? m.emotional_roots : []).map((r) => {
          const d = (r ?? {}) as Dict;
          return { pattern: asStr(d.pattern), note: asStr(d.note) };
        }),
        reframe: asStr(m.reframe),
        inquiryQuestion: asStr(m.inquiry_question),
        somaticPractice: asStr(m.somatic_practice),
        differentialNote: asStr(m.differential_note),
      });
      break; // first phrase that resolves wins; don't double-match a condition
    }
    if (limit && out.length >= limit) break;
  }
  return out;
}

File: fm-database/fmdb/assess/somatic_read.py
"""The chief-complaint read: what the catalogue says about what the client
actually came in for.

WHY THIS AND NOT CONVERGENCE. Convergence ("what do these seven symptoms
share?") only clears its own null model about one time in three, because the
commonest themes — fear in 58% of maps, anger 54%, holding-on 49% — can never
show lift. The most clinically obvious pattern is the one that test is blindest
to.

The chief-complaint read asks a smaller and far more answerable question: for
the thing this client most wants fixed, what does the source say? One map, no
statistics, no threshold — and it fires for every client whose main issue is in
the book. It is what a coach would actually look up.

Matching is alias-aware and tolerant of how conditions are really written in
client.yaml — "Hypertension — ON TREATMENT (previously unreported) — Telma 40"
has to resolve to `hypertension`. A naive slugify finds 2 of Hariharan's maps
where this finds 8.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import Any, Iterable

from .somatic_themes import SomaticTheme, THEME_LABELS, classify_root

#: Words that carry no clinical meaning when matching a condition string.
_NOISE = {
    "suspected", "confirmed", "possible", "probable", "mild", "moderate", "severe",
    "chronic", "acute", "on", "off", "treatment", "previously", "unreported",
    "and", "or", "the", "a", "of", "with", "in", "at", "to", "for", "non", "now",
    "history", "past", "current", "ongoing", "recurrent", "grade", "type",
}


def _phrases(condition: str) -> list[str]:
    """Candidate lookup keys from one free-text condition, longest first.

    Real entries look like "Hypertension — ON TREATMENT (previously unreported)
    — Telma 40 twice daily" or "Type 2 diabetes — HbA1c 6.6% (progressed…)".
    The clinical name is almost always the head, before the first dash, comma,
    colon or bracket.
    """
    head = re.split(r"[—–\-(,:;/]", str(condition))[0]
    head = re.sub(r"\s+", " ", head).strip().lower()
    if not head:
        return []
    words = [w for w in re.findall(r"[a-z0-9']+", head) if w not in _NOISE]
    out: list[str] = []
    if words:
        # BOTH joins, every time. Catalogue aliases are written the way a
        # clinician types them — "underactive thyroid", with a space — while
        # slugs are hyphenated. Emitting only the hyphenated form made 3,130
        # multi-word aliases unreachable, so "Underactive thyroid (subclinical)"
        # matched nothing even though `thyroid-dysfunction` carries exactly that
        # alias. Cheap to ask for both; the resolver ignores what it lacks.
        def _both(ws: list[str]) -> list[str]:
            return ["-".join(ws), " ".join(ws)] if len(ws) > 1 else ["-".join(ws)]

        out.extend(_both(words))                         # full head
        for n in (3, 2):                                 # leading n-grams
            if len(words) > n:
                out.extend(_both(words[:n]))
        if len(words) > 1:
            out.append(words[-1])                        # trailing noun
        out.append(words[0])
    # de-dup, keep order, drop 1-2 char keys which match noisily
    seen, keys = set(), []
    for k in out:
        if len(k) > 2 and k not in seen:
            seen.add(k)
            keys.append(k)
    return keys


@dataclass
class ChiefRead:
    """One matched condition and what the catalogue says about it."""
    condition: str                  # as the coach wrote it
    target_slug: str
    map_slug: str
    display_name: str
    sensitivity: str
    gated: bool                     # coach_only_note set — never auto-surface
    roots: list[tuple[str, str]] = field(default_factory=list)   # (pattern, note)
    themes: list[SomaticTheme] = field(default_factory=list)
    reframe: str = ""
    inquiry_question: str = ""
    somatic_practice: str = ""
    differential_note: str = ""

    @property
    def client_safe(self) -> bool:
        """Safe to surface unsupervised at the client's `full` depth."""
        return self.sensitivity == "general" and not self.gated

    @property
    def theme_labels(self) -> list[str]:
        return [THEME_LABELS[t][0] for t in self.themes]


def read_chief_complaints(
    client: dict[str, Any],
    maps: Iterable[dict[str, Any]],
    resolve: Any,
    limit: int | None = None,
) -> list[ChiefRead]:
    """Map a client's stated conditions onto catalogue somatic maps.

    `resolve(slug) -> canonical slug` is the alias-aware resolver built from the
    live catalogue (symptoms first, then topics) — passed in so this module
    stays free of loader imports and is trivially testable.

    Ordered as the coach wrote the conditions: the first-listed condition is
    almost always the presenting one.
    """
    by_target: dict[str, dict[str, Any]] = {}
    for m in maps:
        by_target.setdefault(resolve(str(m.get("target_slug", ""))), m)

    out: list[ChiefRead] = []
    seen: set[str] = set()
    for cond in (client.get("active_conditions") or []):
        for key in _phrases(str(cond)):
            canon = resolve(key)
            m = by_target.get(canon)
            if not m or m["slug"] in seen:
                continue
            seen.add(m["slug"])
            roots = [
                (str(r.get("pattern", "")), str(r.get("note", "")))
                for r in (m.get("emotional_roots") or [])
            ]
            themes: list[SomaticTheme] = []
            for p, n in roots:
                for t in classify_root(p, n):
                    if t not in themes:
                        themes.append(t)
            out.append(
                ChiefRead(
                    condition=str(cond),
                    target_slug=canon,
                    map_slug=m["slug"],
                    display_name=str(m.get("display_name", "")),
                    sensitivity=str(m.get("sensitivity", "sensitive")),
                    gated=bool(str(m.get("coach_only_note", "")).strip()),
                    roots=roots,
                    themes=themes,
                    reframe=str(m.get("reframe", "")),
                    inquiry_question=str(m.get("inquiry_question", "")),
                    somatic_practice=str(m.get("somatic_practice", "")),
                    differential_note=str(m.get("differential_note", "")),
                )
            )
            break  # first phrase that resolves wins; don't double-match a condition
        if limit and len(out) >= limit:
            break
    return out
